// ya-dns-sd/script/src/mdns/multicast_interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MulticastInterface = void 0;
const message_decode_js_1 = require("../decode/message_decode.js");
const message_encode_js_1 = require("../decode/message_encode.js");
const fast_fifo_js_1 = require("../fast_fifo.js");
const default_driver_node_js_1 = require("./default_driver.node.js");
/** An interface used to send and receive multicast messages, as well as other actions such as toggling multicast loopback.
 *
 * If no driver is specified, selects a `MulticastDriver` appropriate to the current runtime.
 */
class MulticastInterface {
    constructor(driver) {
        Object.defineProperty(this, "driver", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "subscribers", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        const driverToUse = driver || new default_driver_node_js_1.DefaultDriver("IPv4");
        this.driver = driverToUse;
        const subscribers = this.subscribers;
        (async () => {
            while (true) {
                const [received, origin] = await driverToUse.receive();
                try {
                    const event = [(0, message_decode_js_1.decodeMessage)(received), origin];
                    for (const subscriber of subscribers) {
                        subscriber.push(event);
                    }
                }
                catch {
                    console.warn("Could not decode a message from", origin.hostname);
                }
            }
        })();
    }
    send(message) {
        const encoded = (0, message_encode_js_1.encodeMessage)(message);
        return this.driver.send(encoded);
    }
    setTTL(ttl) {
        if (this.driver.family === "IPv4") {
            this.driver.setTTL(ttl);
        }
    }
    setLoopback(loopback) {
        this.driver.setLoopback(loopback);
    }
    messages() {
        const subscriber = new fast_fifo_js_1.FastFIFO(16);
        this.subscribers.push(subscriber);
        return subscriber;
    }
    isOwnAddress(address) {
        return this.driver.isOwnAddress(address);
    }
    get hostname() {
        return this.driver.hostname;
    }
    get family() {
        return this.driver.family;
    }
}
exports.MulticastInterface = MulticastInterface;
